include/common/i2c_interface.h

```diff
@@ -40,25 +40,23 @@ class I2CInterface
     template<typename T, size_t N>
     void read(T (&data)[N])
     {
-        const size_t size = sizeof(T) * N;
-        if (::read(m_I2C, &data[0], size) != size) {
-            throw std::runtime_error("Failed to read from i2c bus");
-        }
+        read(&data[0], N * sizeof(T));
     }
 
+    void read(void *data, size_t size);
+
     void writeByteCommand(uint8_t address, uint8_t byte);
     void writeByte(uint8_t byte);
 
     // writes data
     template<typename T, size_t N>
     void write(const T (&data)[N])
     {
-        const size_t size = sizeof(T) * N;
-        if (::write(m_I2C, &data[0], size) != size) {
-            throw std::runtime_error("Failed to write to i2c bus");
-        }
+        write(&data[0], N * sizeof(T));
     }
 
+    void write(const void *data, size_t size);
```

Comment:
This is a bit of a nasty Cish API. If you wanted to be (unnecessarily?) fancy, you could do:
```c++
template<typename T, 
         typename std::enable_if<std::is_pod<T>::value, T>::type* = nullptr>
void write(const T &data)
{
    if (::write(m_I2C, &data, sizeof(T)) != sizeof(T)) {
        throw std::runtime_error("Failed to write to i2c bus");
    }
}
```